File: 逆向案例/wx小程序之百达屋/bdw.py
import requests
import time
import execjs

obj = execjs.compile(open('bdw.js', encoding='utf-8').read())
message_id = obj.call('getMessageId')
print(message_id)
signature, timestamp = obj.call('getSignature')
print(signature, timestamp)
import requests



headers = {
    'Connection': 'keep-alive',
    'BusinessType': '1',
    'MessageId': message_id,
    'pageSize': '20',
    'pageNo': '2',
    'AppType': '2',
    'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 15_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.0 Mobile/15E148 Safari/604.1 wechatdevtools/1.06.2407110 MicroMessenger/8.0.5 Language/zh_CN webview/',
    'content-type': 'application/json',
    'AppVersion': '3.3.5',
    'OsVersion': 'iOS 10.0.1',
    'DeviceManufacture': 'devtools',
    'Signature': signature,
    'Channel': 'bdw',
    'Timestamp': str(timestamp),
    'ClientType': '5',
    'DeviceModel': 'iPhone 12/13 (Pro)',
    'Accept': '*/*',
    'Sec-Fetch-Site': 'cross-site',
    'Sec-Fetch-Mode': 'cors',
    'Sec-Fetch-Dest': 'empty',
    'Referer': 'https://servicewechat.com/wx62466b68205db6dc/devtools/page-frame.html',
    'Accept-Language': 'zh-CN,zh;q=0.9',
}

json_data = {
    'querySource': 2,
    'channelType': 1,
    'sourceType': 10,
    'serviceType': [
        '3',
    ],
    'searchType': 0,
    'dateStartTime': 1739116800000,
    'dateEndTime': 1739203200000,
    'destination': '国内/国际',
}

response = requests.post('https://api.betterwood.com/base/app/store/listV2', headers=headers, json=json_data)
print(response.json())




# headers = {
#     'Connection': 'keep-alive',
#     'pageNum': '2',
#     'MessageId': message_id,
#     'pageSize': '15',
#     'AppType': '2',
#     'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 15_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.0 Mobile/15E148 Safari/604.1 wechatdevtools/1.06.2407110 MicroMessenger/8.0.5 Language/zh_CN webview/',
#     'content-type': 'application/json',
#     'AppVersion': '3.3.5',
#     'OsVersion': 'iOS 10.0.1',
#     'DeviceManufacture': 'devtools',
#     'Signature': signature,
#     'Channel': 'bdw',
#     'BusinessType': '1',
#     'Timestamp': str(timestamp),
#     'ClientType': '5',
#     'DeviceModel': 'iPhone 12/13 (Pro)',
#     'Accept': '*/*',
#     'Sec-Fetch-Site': 'cross-site',
#     'Sec-Fetch-Mode': 'cors',
#     'Sec-Fetch-Dest': 'empty',
#     'Referer': 'https://servicewechat.com/wx62466b68205db6dc/devtools/page-frame.html',
#     'Accept-Language': 'zh-CN,zh;q=0.9',
#     'Host': 'api.betterwood.com',
#     'Content-Length': '17',
# }
#
# json_data = {
#     'hotelId': '374',
#     'commentType': 0,
#     'roomType': [],
#     'tagId': [],
#     'sortType': 2,
# }
#
# response = requests.post('https://api.betterwood.com/content/hotel/comment/hotel/list', headers=headers, json=json_data)
# print(response.json())

File: 逆向案例/wx小程序之百达屋/bdw.js
const cryptoJS = require('crypto-js');

function c(n) {
    return cryptoJS.MD5(n).toString();
}

function getMessageId() {


    function M(n) {
        for (var e = [8, 13, 18, 23], t = 0; t < e.length; t++)
            n = n.slice(0, e[t]) + "-" + n.slice(e[t]);
        return n;
    }

    function H() {
        for (var n = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : "", e = [], t = "0123456789abcdef", r = 0; r < 36; r++) {
            var o = Math.floor(16 * Math.random());
            e[r] = t.substring(o, o + 1);
        }
        e[14] = "4";
        var u = 3 & Number(e[19]) | 8;
        e[19] = t.substring(u, u + 1),
            e[8] = e[13] = e[18] = e[23] = "-";
        var i = e.join("")
            , a = M(c(i + new Date().getTime() + n));
        return a;
    }

    var r = "";
    I = (0, H)(r)
    return I;
}

function R(n, e, t, r, o, u, i, a, s, f) {
    var d = "987EBBF8450544D7A52D5539DF9A92A2"
        , l = "";
    return l = "AppVersion=".concat(u, "Channel=").concat(i, "ClientType=").concat(o, "DeviceManufacture=").concat(t, "DeviceModel=").concat(r, "MessageId=").concat(a, "OsVersion=").concat(e, "Timestamp=").concat(n, "AppKey=").concat(d, "Url=").concat(f),
        l = c(l.replace(/\s*/g, "")).substring(4, 28).toLocaleUpperCase();
}

function getSignature() {
    var t = {
            AppVersion: "3.3.5",
            Channel: "bdw",
            ClientType: "5",
            DeviceManufacture: "devtools",
            DeviceModel: "iPhone 12/13 (Pro)",
            MessageId: getMessageId(),
            OsVersion: "iOS 10.0.1",
        }
        ,
        o = t.OsVersion
        , u = t.DeviceManufacture
        , i = t.DeviceModel
        , c = t.ClientType
        , a = t.AppVersion
        , f = t.Channel
        , d = t.MessageId
        , p = undefined,
        n = Date.now(),
        url = "https://api.betterwood.com/base/app/store/listV2"
    signature = R(n, o, u, i, c, a, f, d, p, encodeURIComponent(url.split("betterwood.com")[1].split("?")[0]));
    return [signature, n];
}


console.log(getMessageId())
console.log(getSignature())
